lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -169,7 +193,12 @@ def add_payment_method(post, payment_method, options, key = :source)
             post[key][:last_name] = payment_method.last_name if payment_method.last_name
           end
         end
-        unless payment_method.is_a?(String)
+        if payment_method.is_a?(String)
+          if /tok/.match?(payment_method)
+            post[key][:type] = 'token'
```

Comment:
💬 Comment:

Not sure if the proper behavior is to name space this post[:source][:type] when you are going to remove the `source` level on the `store` method, should be simpler just to put `type` and `token` at the root level of the post object ?

❓ Question:

I can uderstand why you need to move `type` and `token`  one level up, but it aslo applies for billing address ?

```ruby
### Before
{:source=>
  {:type=>"token",
   :token=>"tok_nlnv2zsawtbulj577oxrzaa7pe",
   :billing_address=>
    {:address_line1=>"456 My Street",
     :address_line2=>"Apt 1",
     :city=>"Ottawa",
     :state=>"ON",
     :country=>"CA",
     :zip=>"K1C2N6"}},
 :customer=>{:email=>"longbob.longsen@example.com"}}

## After
{:customer=>{:email=>"longbob.longsen@example.com"},
 :type=>"token",
 :token=>"tok_nlnv2zsawtbulj577oxrzaa7pe",
 :billing_address=>
  {:address_line1=>"456 My Street",
   :address_line2=>"Apt 1",
   :city=>"Ottawa",
   :state=>"ON",
   :country=>"CA",
   :zip=>"K1C2N6"}}
```

Reply:
This example doesn't apply because, after the tokenizing process, the add_customer_data only will add the :customer=>{:email=>"longbob.longsen@example.com"}. The reason the add_customer_data requires the existence of post[:source] to add billing_address information. So I will need to use add_payment_method, which also adds the source hash, and add to this source hash all the billing address information through the add_customer_data, and then move the type and token to one level up and with that the billing address, at the end still working in this way.



```ruby
# token request
### before
{:source=>
  {:type=>"card",
   :name=>"Longbob Longsen",
   :number=>"4242424242424242",
   :cvv=>"100",
   :first_name=>"Longbob",
   :last_name=>"Longsen",
   :expiry_year=>"2025",
   :expiry_month=>"06",
   :billing_address=>{:address_line1=>"456 My Street", :address_line2=>"Apt 1", :city=>"Ottawa", :state=>"ON", :country=>"CA", :zip=>"K1C2N6"}},
 :customer=>{:email=>"longbob.longsen@example.com"}}

### after
{:type=>"card",
 :name=>"Longbob Longsen",
 :number=>"4242424242424242",
 :cvv=>"100",
 :first_name=>"Longbob",
 :last_name=>"Longsen",
 :expiry_year=>"2025",
 :expiry_month=>"06",
 :billing_address=>{:address_line1=>"456 My Street", :address_line2=>"Apt 1", :city=>"Ottawa", :state=>"ON", :country=>"CA", :zip=>"K1C2N6"}}
```

```
### store request
{:type=>"token", :token=>"tok_fytopfhgnimuxb753662nzymre", :customer=>{:email=>"longbob.longsen@example.com"}}
``
`
We don't need to add more billing_address values because the stored card has the information, and the request-response contains the customer_id and email associated.